[Chat] 채팅방 응답에 사용자 정보 추가 및 DTO 구조 개선

yunjeooong opened the pull request:
# [Chat] 채팅방 응답에 사용자 정보 추가 및 DTO 구조 개선

## Description
### 1. 이 일이 왜 시작되었는가?
- 채팅방 목록과 상세 조회 시 사용자의 프로필 사진과 닉네임 정보가 필요
- 채팅방에서 한 사용자가 나갔을 때 이를 구분하고 적절한 메시지를 표시해야 함
- 채팅방 나가기 후 메시지 전송 권한 관리 필요
-  ChatRoomController에 Swagger API 문서화 추가 필요


### 2. 배경 설명
- 기존에는 채팅방 상태가 단순히 ACTIVE/INACTIVE로만 구분
- 누가 채팅방을 나갔는지 구분할 수 없었음

## Changes

### 1. 채팅방 상태 관리 개선
- `ChatRoomLeftBy` enum 추가
  - NONE, BUYER, SELLER 상태 구분
  - 각 상태별 메시지 정의
- `ChatRoom` 엔티티에 `leftBy` 필드 추가
- 채팅방 나가기 관련 메서드 개선
  - `hasUserLeft`: 사용자의 퇴장 여부 확인
  - `canSendMessage`: 메시지 전송 권한 확인
  - `getExitMessage`: 퇴장 메시지 반환

### 2. ChatRoomResponseDTO 구조 개선
- 도메인 개념에 맞춰 계층적으로 재구성
  ```java
  public record ChatRoomResponseDTO(
      ChatRoomInfoDTO info,          // 채팅방 기본 정보
      ParticipantsDTO participants,  // 참여자 정보
      ChatRoomStatusDTO status       // 채팅방 상태 정보
  )
  ```
- 각 하위 DTO의 역할
  - `ChatRoomInfoDTO`: id, roomId, articleId, createdAt
  - `ParticipantsDTO`: buyer와 seller의 상세 정보(닉네임, 프로필 이미지)
  - `ChatRoomStatusDTO`: 메시지 전송 가능 여부, 퇴장 메시지

### 3. 채팅방 조회 로직 수정
- 채팅방 목록 조회 시 나간 사용자 제외
- 채팅방 상세 조회 시 접근 권한 체크 강화
- 사용자 정보(닉네임, 프로필 이미지) 포함하여 응답

yunjeooong committed:
feat:구매자와 판매자의 닉네임과 프로필 이미지 필드를 추가

yunjeooong committed:
feat:feat: ChatRoomController에 Swagger API 문서화 추가

yunjeooong committed:
feat: 채팅방 나가기 기능 개선 및 상태 관리 방식 변경

- 채팅방 나가기 상태 추적을 위한 ChatRoomLeftBy enum 추가
- 채팅방 엔티티에 퇴장 사용자 구분 필드(leftBy) 추가
- 나간 사용자의 메시지 전송 및 조회 권한 제한 기능 추가

yunjeooong committed:
refactor: ChatRoomResponseDTO 구조 개선

- ChatRoomResponseDTO를 도메인 개념에 맞춰 계층적으로 재구성
  - ChatRoomInfoDTO: 채팅방 기본 정보
  - ParticipantsDTO: 구매자/판매자 정보
  - ChatRoomStatusDTO: 채팅방 상태 정보

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/ChatRoomService.java:
```diff
@@ -44,21 +47,32 @@ public ChatRoomResponseDTO createChatRoom(ChatRoomRequestDTO requestDto) {
-        List<ChatRoom> sellerRooms = chatRoomRepository
-                .findByStatusAndSellerIdOrderByCreatedAtDesc(ChatRoomStatus.ACTIVE, userId);
+        // 구매자 또는 판매자로 참여한 모든 채팅방을 가져옴
+        List<ChatRoom> buyerRooms = chatRoomRepository.findByBuyerIdOrderByCreatedAtDesc(userId);
```
너무 비효율적이지 않을까용 
그냥 사용자가 참여한 모든 채팅방을 가져오는게 더 나아보이는데...

yunjeooong replied:
규현님 항상 꼼꼼한 리뷰 정말 감사합니다 🙏
규현님 말씀처럼 기존에는  불필요한 DB 접근이 많았습니다🥹

before:
1. 채팅방 조회 시 불필요한 분리 조회
 ``` java
  List<ChatRoom> buyerRooms = chatRoomRepository.findByBuyerIdOrderByCreatedAtDesc(userId);
  List<ChatRoom> sellerRooms = chatRoomRepository.findBySellerIdOrderByCreatedAtDesc(userId);

```
2. N+1 문제 발생
 ``` java
  .map(chatRoom -> {
       User buyer = userRepository.findById(chatRoom.getBuyerId())...
       User seller = userRepository.findById(chatRoom.getSellerId())...
   })
```
각 채팅방마다 구매자와 판매자 정보를 별도로 조회하게 되어채팅방 10개만 있어도 총 22번의 DB 쿼리 발생
채팅방 조회: 2회
사용자 정보 조회: 20회 (채팅방당 2회) ...
와 같은 문제점이 있었슴니다.,

다음을 개선하고자 아래 내용과 같이 개선했습니당!

👉🏻After:
1. 단일 쿼리 개선
``` java
   @Query("""
       SELECT DISTINCT cr FROM ChatRoom cr
       JOIN FETCH User buyer ON buyer.userId = cr.buyerId
       JOIN FETCH User seller ON seller.userId = cr.sellerId
       WHERE cr.buyerId = :userId OR cr.sellerId = :userId
       ORDER BY cr.createdAt DESC
       """)
```
- JOIN FETCH를 활용하여 채팅방과 사용자 정보를 한 번에 조회
- 구매자/판매자 구분 없이 통합 조회
- 정확한 시간순 정렬 보장

2. DB 접근 최소화

- 채팅방 + 사용자 정보 조회: 1회
- MongoDB 메타 정보 조회: 1회
- 채팅방 수와 관계없이 항상 2회의 쿼리만 발생

3. 메모리 사용 최적화
``` java
   return chatRooms.stream()
           .filter(chatRoom -> !chatRoom.hasUserLeft(userId))
           .map(chatRoom -> {
               ChatroomMetaInfo metaInfo = metaInfoMap.get(chatRoom.getRoomId());
               return ChatRoomResponseDTO.of(
                       chatRoom,
                       chatRoom.getBuyer(),  // 이미 조회된 정보 사용
                       chatRoom.getSeller(), // 이미 조회된 정보 사용
                       userId
               );
           })
```

- 이미 조회된 사용자 정보를 재사용

**📌 결론적으로!!!!**
채팅방 10개 기준
Before: 22회 쿼리 → After: 2회 쿼리
약 91% 쿼리 감소가 되도록 개선하였습니다

다시한번 리뷰 감사드립니다!!!

beetrbgus replied:
정성적인 PR 코멘트 감사합니당-

beetrbgus commented on src/main/java/com/ani/taku_backend/chatroom/service/ChatRoomService.java:
```diff
@@ -75,11 +89,21 @@ public ChatRoomResponseDTO findChatRoom(String roomId, Long userId) {
         }
 
-        return ChatRoomResponseDTO.of(chatRoom);
+        User buyer = userRepository.findById(chatRoom.getBuyerId())
```
이부분에서도 마찬가지로 너무 DB 접속을 많이 유지를 하는 것 같아요. 왜 판매자와 구매자를 나눠서 조회하나요!?

yunjeooong committed:
fix: 채팅방 조회 성능 개선을 위한 N+1 문제 해결

- ChatRoom 엔티티에 buyer, seller 필드 추가
  - @ManyToOne(fetch = FetchType.LAZY) 설정으로 지연 로딩 적용
  - insertable=false, updatable=false로 설정하여 데이터 정합성 보장

- ChatRoomRepository 수정
  - findAllByUserIdWithParticipants: JOIN FETCH를 사용하여 채팅방과 참여자 정보를 단일 쿼리로 조회
  - findByRoomIdWithParticipants: 채팅방 상세 조회 시에도 JOIN FETCH 적용

- ChatRoomService 최적화
  - findChatRoomList: N+2 쿼리를 2개 쿼리로 감소 (채팅방+참여자 조회 1회, 메타정보 조회 1회)
  - findChatRoom: 추가 쿼리 없이 참여자 정보 즉시 사용 가능하도록 개선

yunjeooong committed:
chore: 미사용 메서드에 대한 주석 처리 추가

Baejjjjjjjj commented on src/main/java/com/ani/taku_backend/chatroom/model/constant/ChatRoomLeftBy.java:
```diff
@@ -0,0 +1,23 @@
+public enum ChatRoomLeftBy {
+    NONE("없음", "아무도 나가지 않았습니다."),
+    BUYER("구매자", "구매자가 채팅방을 나갔습니다."),
+    SELLER("판매자", "판매자가 채팅방을 나갔습니다.");
```
혹시 구매자와 판매자가 모두 나갔을 때는 이 값을 어떻게 설정해야할까요?
